src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1024,3 +1073,480 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
+        table: str,
+        *,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        max_workers: Optional[int] = None,
+        query_timeout: Optional[int] = 0,
+        fetch_size: Optional[int] = 0,
+        custom_schema: Optional[Union[str, StructType]] = None,
+        predicates: Optional[List[str]] = None,
+        session_init_statement: Optional[str] = None,
+    ) -> DataFrame:
+        """Reads data from a database table using a DBAPI connection.
+        Args:
+            create_connection: a function that return a dbapi connection
+            table: the name of the table in external data source
+            column: column name used to create partition
+            lower_bound: lower bound of partition
+            upper_bound: upper bound of partition
+            num_partitions: number of partitions to create
+            max_workers: number of workers for parallelism
+            query_timeout: timeout for each query, default value is 0, meaning never timeout
+            fetch_size: batch size when fetching from external data source
+            custom_schema: a custom snowflake table schema to read data from external data source, the column names should be identical to corresponded column names external data source
+            predicates: a list of expressions suitable for inclusion in WHERE clauses, each defines a partition
+            session_init_statement: session initiation statements for external data source
+        Note:
+            column, lower_bound, upper_bound and num_partitions must be specified if any one of them is specified.
+        """
+        statements_params_for_telemetry = {STATEMENT_PARAMS_DATA_SOURCE: "1"}
+        start_time = time.perf_counter()
+        conn = create_connection()
+        dbms_type, driver_info = detect_dbms(conn)
+        logger.info(f"Detected DBMS: {dbms_type}, Driver Info: {driver_info}")
+        if custom_schema is None:
+            struct_schema = infer_data_source_schema(
+                conn, table, dbms_type, driver_info
+            )
+        else:
+            if isinstance(custom_schema, str):
+                struct_schema = type_string_to_type_object(custom_schema)
+                if not isinstance(struct_schema, StructType):
+                    logger.error(f"Invalid schema string: {custom_schema}")
+                    raise ValueError(
+                        f"Invalid schema string: {custom_schema}. "
+                        f"You should provide a valid schema string representing a struct type."
+                    )
+            elif isinstance(custom_schema, StructType):
+                struct_schema = custom_schema
+            else:
+                logger.error(f"Invalid schema type: {type(custom_schema)}")
+                raise TypeError(f"Invalid schema type: {type(custom_schema)}.")
```

Comment:
TypeError is usually for Python Data Type error.
Better give users instructions how to use it.

Reply:
updated